joblib/_store_backends.py

```diff
@@ -185,12 +186,24 @@ def dump_item(self, path, item, verbose=1):
 
             def write_func(to_write, dest_filename):
                 with self._open_item(dest_filename, "wb") as f:
-                    numpy_pickle.dump(to_write, f,
-                                      compress=self.compress)
+                    try:
+                        numpy_pickle.dump(to_write, f, compress=self.compress)
+                    except PicklingError as e:
+                        warnings.warn(
+                            'Unable to cache to disk: '
+                            'failed to pickle output. '
+                            'In future versions of joblib '
+                            'this will raise an exception.'
+                            'Exception: %s.' % e,
+                            FutureWarning
+                        )
 
             self._concurrency_safe_write(item, filename, write_func)
-        except:  # noqa: E722
-            " Race condition in the creation of the directory "
+        except Exception as e:  # noqa: E722
+            warnings.warn(
+                'Unable to cache to disk. '
+                'Possibly a race condition in the creation of the directory. '
+                'Exception: %s' % e)
```

Comment:
As we are changing a bit this, I would be in favor of sanitizing this, by putting a `try/except` for `OSError` in `create_location`, and having this raise a deprecation warning saying that any other error will be raised in future version of `joblib`. WDYT?


Small nitpick on style.
```suggestion
        except Exception as e:  # noqa: E722
            warnings.warn(
                'Unable to cache to disk. '
                'Possibly a race condition in the creation of the directory. '
                f'Exception: {e}'
            )
```

Reply:
So following the conversation with @ogrisel , I would not do a deprecation warning but just a warning for the error.
Maybe create a specfic `CacheWarning` that will be easy to filter out?